### ISSUE TITLE
verifytypes on class with property incorrectly says that type is partially unknown

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#1489]
**Describe the bug**
When creating a class variable using something like `name=property(fset=..., fget=...)` and `pyright --verifytypes`, pyright reports that the property has a partially unknown type

**To Reproduce**
Create a library `foolib` with empty `__init__.py` and `py.typed` files.
Add the following as `fooclass.py`
```python
class Foo:
    """A test class"""

    def __init__(self):
        self._name: str = ""
        pass

    def _set_name(self, name: str):
        self._name = name

    def _get_name(self) -> str:
        return self._name

    name = property(
        fset=_set_name,
        fget=_get_name,
    )
```

`pyright --verifytypes foolib` reports
```
Package name: "foolib"
Package directory: "foolib"
Path of py.typed file: "foolib\py.typed"
  error: Type partially unknown for class "foolib.fooclass.Foo"
    Type partially unknown for class "Foo"
      Type unknown for variable "name"
  error: Type not declared for variable "foolib.fooclass.Foo.name"

```

**Expected behavior**
No error

**VS Code extension or command-line**
command line:  pyright version 1.1.110

**Additional context**
Came up when trying to clean up some pandas stuff

**Comment by erictraut** [comment]
Properties require significant special-case code within a type checker. I know that it's possible to define a property in ways other than the typical decorators, but I've very rarely run across it in practice. I'd prefer not to add more special-case code paths to handle this rare case, especially when there's a code workaround that is arguably better. Is there a good reason why the standard `@property` and `@x.setter` decorators won't work in this case?

**Comment by Dr-Irv** [comment]
I created this simple example from a more complex example in the pandas code.

I think that the issue occurs when the setter can take one kind of argument (e.g., `Union[str, List[str]]`) but the getter always returns a specific type, e.g. `List[str]` .  If you use `@property` in this case, you get other complaints (if I remember correctly from some other code I was working with).  Here's an example:
```python
    @property
    def aprop(self) -> List[str]:
        return self._aprop

    @aprop.setter
    def _set_aprop(self, theprop: Union[str, List[str]]):
        if isinstance(theprop, str):
            lprop = [theprop]
        else:
            lprop = theprop
        self._aprop = lprop
```
Here on `set_aprop`, you will get the message:
```
Property setter value type is not assignable to the getter return type
  Type "str | List[str]" cannot be assigned to type "List[str]"
    "str" is incompatible with "List[str]"
```

In the case of pandas, I don't think I'd get the core contributors to agree to remove the use of `= property` just to satisfy the type checker.

**Comment by erictraut** [comment]
Pyright reports inconsistencies between getters and setters only in "strict" mode, so this shouldn't appear in normal type checking mode.

So you're saying that the justification for defining this property in a non-standard manner is to avoid an error from a type checker? But you're also saying that the maintainers probably wouldn't be willing to make a recommended change to satisfy a type checker? That sounds inconsistent to me. There must be some other reason why it was defined in a non-standard way. Or maybe there wasn't a good reason behind it.

**Comment by erictraut** [comment]
The purpose of the "type completeness" report is not only to ensure that all types are specified but also that all type checkers will evaluate the same types. I don't think we can rely on all type checkers supporting this non-standard form of property instantiation. Therefore, I think it would run counter to the goal of the type verification feature to address this issue. Please try to use the standard property decorators which all type checkers will interpret correctly.

**Comment by Dr-Irv** [comment]
> So you're saying that the justification for defining this property in a non-standard manner is to avoid an error from a type checker? But you're also saying that the maintainers probably wouldn't be willing to make a recommended change to satisfy a type checker? That sounds inconsistent to me. There must be some other reason why it was defined in a non-standard way. Or maybe there wasn't a good reason behind it.

The other issue is that we use `mypy` to verify our code, and it complains when the `setter` and `getter` don't have matching types.  See discussion https://github.com/python/mypy/issues/3004 and it seems this has been an issue before for `pyright`: https://github.com/microsoft/pyright/issues/993

Is it possible for you to add a command line option that ignores this?  Or is there a `# type [ignore]` thing that would work?

**Comment by erictraut** [comment]
Yeah, you can use a `# type: ignore` to suppress the error that mypy emits here. This is used in a couple of places in the typeshed stdlib stubs where a getter and setter type don't match.
